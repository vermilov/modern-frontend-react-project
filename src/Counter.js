import React from "react";

export const Counter = ({ counter }) => (
    <span>
        <p>{counter}</p>
    </span>
);

export default Counter;
