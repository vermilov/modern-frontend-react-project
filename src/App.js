// import React, { useReducer, createContext } from "react";
// import uuid from "uuid/v4";
// import Filter from "./Filter";
// import TodoList from "./TodoList";
// import AddTodo from "./AddTodo";

// export const TodoContext = createContext(null);

// const initialTodos = [
//     {
//         id: uuid(),
//         task: "Learn React",
//         complete: true,
//     },
//     {
//         id: uuid(),
//         task: "Learn Svelte",
//         complete: true,
//     },
//     {
//         id: uuid(),
//         task: "Learn MySQL",
//         complete: false,
//     },
// ];

// const filterReducer = (state, action) => {
//     switch (action.type) {
//         case "SHOW_ALL":
//             return "ALL";
//         case "SHOW_COMPLETE":
//             return "COMPLETE";
//         case "SHOW_INCOMPLETE":
//             return "INCOMPLETE";
//         default:
//             throw new Exception();
//     }
// };

// const todoReducer = (state, action) => {
//     switch (action.type) {
//         case "DO_TODO": {
//             return state.map(todo => {
//                 if (todo.id === action.id) {
//                     return { ...todo, complete: true };
//                 } else {
//                     return todo;
//                 }
//             });
//         }
//         case "UNDO_TODO": {
//             return state.map(todo => {
//                 if (todo.id === action.id) {
//                     return { ...todo, complete: false };
//                 } else {
//                     return todo;
//                 }
//             });
//         }
//         case "ADD_TODO": {
//             return [
//                 ...state,
//                 { id: action.id, task: action.task, complete: false },
//             ];
//         }
//         default: {
//             throw new Error();
//         }
//     }
// };

// const App = () => {
//     const [filter, dispatchFilter] = useReducer(filterReducer, "ALL");
//     const [todos, dispatchTodos] = useReducer(
//         todoReducer,
//         initialTodos,
//     );

//     const filteredToDos = todos.filter(todo => {
//         if (filter === "ALL") {
//             return true;
//         }

//         if (filter === "COMPLETE" && todo.complete) {
//             return true;
//         }

//         if (filter === "INCOMPLETE" && !todo.complete) {
//             return true;
//         }

//         return false;
//     });
//     return (
//         <TodoContext.Provider value={dispatchTodos}>
//             <Filter dispatch={dispatchFilter} />
//             <TodoList todos={filteredToDos} />
//             <AddTodo />
//         </TodoContext.Provider>
//     );
// };

import React from "react";
import axios from "axios";
import Counter from "./Counter";

export const dataReducer = (state, action) => {
    if (action.type === "SET_ERROR") {
        return { ...state, list: [], error: true };
    }

    if (action.type === "SET_LIST") {
        return { ...state, list: action.list, error: null };
    }

    throw new Error();
};

const initialData = {
    list: [],
    error: null,
};

const App = () => {
    const [counter, setCounter] = React.useState(0);
    const [data, dispatch] = React.useReducer(
        dataReducer,
        initialData,
    );

    React.useEffect(() => {
        axios
            .get("http://hn.algolia.com/api/v1/search?query=react")
            .then(response => {
                dispatch({
                    type: "SET_LIST",
                    list: response.data.hits,
                });
            })
            .catch(() => {
                dispatch({ type: "SET_ERROR" });
            });
    }, []);

    return (
        <div>
            <h1>My Counter</h1>
            <Counter counter={counter} />

            <button
                type="button"
                onClick={() => setCounter(counter + 1)}
            >
                Increment
            </button>

            <button
                type="button"
                onClick={() => setCounter(counter - 1)}
            >
                Decrement
            </button>

            <h2>My Async Task</h2>

            {data.error && <div className="error">Error</div>}

            <ul>
                {data.list.map(item => (
                    <li key={item.objectID}>{item.title}</li>
                ))}
            </ul>
        </div>
    );
};

export default App;
