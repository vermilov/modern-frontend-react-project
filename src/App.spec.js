import React from "react";
import renderer from "react-test-renderer";
import { mount } from "enzyme";
import App from "./App";
import Counter from "./Counter";
import axios from "axios";

describe("App tests", () => {
    test("Snapshot renderers", () => {
        const component = renderer.create(<App />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("Renders the inner Counter", () => {
        const wrapper = mount(<App />);
        expect(wrapper.find(Counter).length).toEqual(1);
    });

    it("Passes all props to counter", () => {
        const wrapper = mount(<App />);
        const counter = wrapper.find(Counter);

        expect(counter.find("p").text()).toEqual("0");
    });

    it("Increments the counter", () => {
        const wrapper = mount(<App />);
        wrapper
            .find("button")
            .at(0)
            .simulate("click");

        const counter = wrapper.find(Counter);
        expect(counter.find("p").text()).toEqual("1");
    });

    it("Decrements the counter", () => {
        const wrapper = mount(<App />);
        wrapper
            .find("button")
            .at(1)
            .simulate("click");

        const counter = wrapper.find(Counter);
        expect(counter.find("p").text()).toEqual("-1");
    });

    it("Fetches async data", done => {
        const promise = new Promise((resolve, reject) => {
            setTimeout(
                () =>
                    resolve({
                        data: {
                            hits: [
                                { objectID: "1", title: "a" },
                                { objectID: "2", title: "b" },
                            ],
                        },
                    }),
                100,
            );
        });

        axios.get = jest.fn(() => promise);

        const wrapper = mount(<App />);

        expect(wrapper.find("li").length).toEqual(0);

        promise.then(() => {
            setImmediate(() => {
                wrapper.update();

                expect(wrapper.find("li").length).toEqual(2);

                axios.get.mockClear();

                done();
            });
        });
    });

    it("Fetches async data but fails", done => {
        const promise = new Promise((resolve, reject) =>
            setTimeout(() => reject(new Error("Whoops!")), 100),
        );

        axios.get = jest.fn(() => promise);

        const wrapper = mount(<App />);

        expect(wrapper.find("li").length).toEqual(0);

        promise.catch(() => {
            setImmediate(() => {
                wrapper.update();

                expect(wrapper.find("li").length).toEqual(0);
                expect(wrapper.find(".error").length).toEqual(1);

                axios.get.mockClear();

                done();
            });
        });
    });
});
