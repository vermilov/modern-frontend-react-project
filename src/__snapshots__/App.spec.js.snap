// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App tests Snapshot renderers 1`] = `
<div>
  <h1>
    My Counter
  </h1>
  <span>
    <p>
      0
    </p>
  </span>
  <button
    onClick={[Function]}
    type="button"
  >
    Increment
  </button>
  <button
    onClick={[Function]}
    type="button"
  >
    Decrement
  </button>
  <h2>
    My Async Task
  </h2>
  <ul />
</div>
`;
